### PR TITLE
[FIRRTL] Support 'module' as identifier for instance names

### LINKED ISSUE #4564: [FIRRTL] Parsing broken for "an instance called module"
FIRRTL parsing is broken for the following circuit:

```
circuit Foo:
  module Bar:
    skip

  module Foo:

    inst module of Bar
```

The parser thinks that `module` has to be the first token when it can be an identifier:

```
build/Bar.fir:7:10: error: 'module' must be first token on its line
    inst module of Bar
         ^
```

h/t @chick for finding this one.

### PR DESCRIPTION
**Pull request by adkian-sifive**
Detect the 'module' string being parsed as `kw_module` instead of `identifier` when it's used as an identifier for instance names. 

Fixes #4564 and #3363

### CODE REVIEW

**Comment by trilorez** [comment]
I think this can be handled cleaner without adding contextual state to the lexer.

Module parsing is deferred until all module symbols are known. The error seems to be in the logic to determine when one is onto the next module to parse. If you look [here](https://github.com/llvm/circt/blob/main/lib/Dialect/FIRRTL/Import/FIRParser.cpp#L2950), the parser assumes any instance of a `module` or `extmodule` keyword means a new module has started. It should really check that `module` or `extmodule` is the first token of a new line. You can check this with `getIndention()` not being `nullopt`.

With this approach, both `module` and `extmodule` should work as instance identifiers automatically.

**Comment by adkian-sifive** [comment]
Would that still fail this check: https://github.com/llvm/circt/blob/48bac7a58f0c85e94073791dadc92ad72d0ccd33/lib/Dialect/FIRRTL/Import/FIRParser.cpp#L3209 since the token kind is still `kw_module`?

**Comment by trilorez** [comment]
That check happens when parsing at the top level and shouldn't affect anything. When the instance expression gets parsed, it will look for an identifier right after the `inst` keyword [here](https://github.com/llvm/circt/blob/main/lib/Dialect/FIRRTL/Import/FIRParser.cpp#L2290). And like I linked above, `parseId` already accepts keywords.

**Comment by trilorez** [comment]
Hm, you'll also need to make sure actually accessing the instance works properly. For example, `module.member <= UInt(1)`. It might be sufficient for the indentation check I suggested above to also see if the indentation level is within a module body but I'm not sure.

**Comment by youngar** [comment]
I agree with @trilorez. I don't really know how whitespace sensitive grammars work in general but maybe the lexer could return `indent` and `dedent` tokens. Might be interesting if you could find a way to tackle bugs like https://github.com/llvm/circt/issues/3273 or https://github.com/llvm/circt/issues/2978.

**Comment by richardxia** [comment]
> I don't really know how whitespace sensitive grammars work in general but maybe the lexer could return `indent` and `dedent` tokens.

You are much more of a compiler expert than I am, but I am very familiar with Python, which has significant whitespace, and they [do indeed emit indent and dedent tokens](https://docs.python.org/3/reference/lexical_analysis.html#indentation) with their special lexer that keeps track of the indentation level.

**Comment by dtzSiFive** [comment]
FWIW module keyword can start beyond column 2, FIRRTL is flexible in how many spaces are used (can be 1 or higher), with rules about consistency:

https://github.com/chipsalliance/firrtl-spec/blob/main/spec.md#details-about-syntax

As @youngar points out, since modules must be same indent level can probably take advantage of that?

But I think the latest commit checking if indentation is at most 2 is not compatible with spec presently.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -2946,9 +2947,13 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit,
...
         // If we got to the next module, then we're done.
       case FIRToken::kw_module:
-      case FIRToken::kw_extmodule:
-        return success();
-
+      case FIRToken::kw_extmodule: {
+        // If the module is not indented to at most 2 it's actually an
+        // identifier
+        auto indent = getIndentation();
+        if (indent.has_value() && indent <= moduleIndent)
```
Comment (trilorez) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2954]: As @dtzSiFive pointed out, hardcoding the width amount here goes against the spec. I think you can just grab `getIdentation` before this `while` loop and compare the `getIndentation` here to that.

File: test/firtool/module-identifier.fir [hunk 2]
```diff
@@ -0,0 +1,14 @@
+; RUN: firtool %s --format=fir --parse-only -verify-diagnostics
```
Comment (uenoku) [inline test/firtool/module-identifier.fir:1]: I think `FileCheck` is missing. 
```suggestion
; RUN: firtool %s --format=fir --parse-only -verify-diagnostics | FileCheck %s
```

**Review by uenoku — COMMENTED** [review]
nit: Probably we might want to move the test to `test/Dialect/FIRRTL/parse-basic.fir`

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 3]
```diff
@@ -224,8 +224,7 @@ struct FIRParser {
                               const Twine &message);
 
   // Parse ('<' intLit '>')? setting result to -1 if not present.
-  template <typename T>
-  ParseResult parseOptionalWidth(T &result);
+  template <typename T> ParseResult parseOptionalWidth(T &result);
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:227]: I think this is violating llvm style guidelines (80 character lines).

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 4]
```diff
@@ -2765,6 +2763,9 @@ struct FIRCircuitParser : public FIRParser {
 
   SmallVector<DeferredModuleToParse, 0> deferredModules;
   ModuleOp mlirModule;
+
+  // Indentation level for 'module' when not used as an identifier
+  const unsigned moduleIndent = 2;
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2768]: There shouldn't be a hard coded indent.  I believe firrtl allows 1 space indent too, or 3, etc.

File: test/firtool/module-identifier.fir [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+; Test parsing of 'module' as an identifier for instance names
+
+; CHECK-LABEL: moduleIdentifier
```
Comment (darthscsi) [inline test/firtool/module-identifier.fir:5]: Please add actual diagnostics.

File: test/firtool/module-identifier.fir (hunk 2 above)
Comment (adkian-sifive) [reply test/firtool/module-identifier.fir]: Moved test to parse-basic.fir

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 6]
```diff
@@ -2946,9 +2946,13 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit,
...
-        return success();
-
+      case FIRToken::kw_extmodule: {
+        // All module declerations should have the same indentation
+        // level.  Use this fact to differentiate between module
+        // declarations and usages of "module" as identifiers.
+        if (getIndentation() == indent)
+          return success();
+      }
       default:
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2956]:
```suggestion
        [[fallthrough]];
      default:
```

File: test/Dialect/FIRRTL/parse-basic.fir [hunk 7]
```diff
@@ -1113,3 +1113,17 @@ circuit MyModule :     ; CHECK: firrtl.circuit "MyModule" {
...
+    out <= in
+
+    ; Module as identifier
+    ; Test parsing of "module" as an identifier for instance names
+
+  module ModuleAsIdentifier:
+    inst module of SomeModule
+    ; CHECK: %module_in, %module_out = firrtl.instance module
+    module.in <= UInt(1)
+    ; CHECK: firrtl.connect %module_in, %c1_ui : !firrtl.uint, !firrtl.uint
```
Comment (seldridge) [inline test/Dialect/FIRRTL/parse-basic.fir:1129]: Mega-nit: This test can be slimmed down to just:

```
module SomeModule:

module ModuleAsIdentifier:
  inst module of SomeModule
```

**Review by seldridge — APPROVED** [review]

File: test/Dialect/FIRRTL/parse-basic.fir (hunk 7 above)
Comment (adkian-sifive) [reply test/Dialect/FIRRTL/parse-basic.fir:1129]: I think I added the rest to check for accessing the instance as Daniel suggested here https://github.com/llvm/circt/pull/4589#issuecomment-1404155876

Comment (adkian-sifive) [reply test/Dialect/FIRRTL/parse-basic.fir:1129]: Not sure if that's actually necessary..?

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 8]
```diff
@@ -2946,9 +2946,13 @@ ParseResult FIRCircuitParser::parseModule(CircuitOp circuit,
...
         // If we got to the next module, then we're done.
       case FIRToken::kw_module:
-      case FIRToken::kw_extmodule:
-        return success();
-
+      case FIRToken::kw_extmodule: {
+        // All module declerations should have the same indentation
+        // level.  Use this fact to differentiate between module
+        // declarations and usages of "module" as identifiers.
+        if (getIndentation() == indent)
```
Comment (dtzSiFive) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2953]: Note, no action requested:
Presently, the test input with one less space before `module.in` is not rejected, and ignored entirely if drops below "module" level.  Both are something the module parser probably can/should reject (elsewhere, separate PR/effort), although could reject here I believe if less-than the indentation of the other modules?

File: test/Dialect/FIRRTL/parse-basic.fir (hunk 7 above)
Comment (dtzSiFive) [reply test/Dialect/FIRRTL/parse-basic.fir:1129]: Knocking test down a little and using `module` as name of a module too while at it:
```mlir
  ; Module as identifier
  ; Test parsing of "module" as an identifier for instance and module names
  extmodule module:
    input in: UInt

  module ModuleAsIdentifier:
    inst module of module
    ; CHECK: firrtl.instance module interesting_name @module
    module.in <= UInt(1)
```

Leaving the instance test for now, as looks like there's work to be done there (see other comment).

**Comment by dtzSiFive** [comment]
Since the issue mentioned above (no longer rejecting statements with "module" if at wrong indentation level) is really the same issue we already have and have issues for, I think this is good to go with the suggested test simplification.

**Review by dtzSiFive — APPROVED** [review]
LGTM, thanks!

**Pull request merged**